src/MongoDB/TopologyDescription.c

```diff
@@ -0,0 +1,248 @@
+#endif
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+
+zend_class_entry* php_phongo_topologydescription_ce;
+
+#define PHONGO_TOPOLOGY_UNKNOWN "Unknown"
+#define PHONGO_TOPOLOGY_SINGLE "Single"
+#define PHONGO_TOPOLOGY_SHARDED "Sharded"
+#define PHONGO_TOPOLOGY_REPLICA_SET_NO_PRIMARY "ReplicaSetNoPrimary"
+#define PHONGO_TOPOLOGY_REPLICA_SET_WITH_PRIMARY "ReplicaSetWithPrimary"
+
+/* {{{ proto array MongoDB\Driver\TopologyDescription::getServers()
+   Returns an array of ServerDescription objects for all known servers in the topology */
+static PHP_METHOD(TopologyDescription, getServers)
+{
+	php_phongo_topologydescription_t* intern;
+	mongoc_server_description_t**     sds;
+	size_t                            i, n = 0;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	sds = mongoc_topology_description_get_servers(intern->topology_description, &n);
+
+	array_init_size(return_value, n);
+
+	for (i = 0; i < n; i++) {
+		zval obj;
+
+		phongo_serverdescription_init(&obj, sds[i]);
+		add_next_index_zval(return_value, &obj);
+	}
+} /* }}} */
+
+/* {{{ proto boolean MongoDB\Driver\TopologyDescription::hasReadableServer([?MongoDB\Driver\ReadPreference $readPreference])
+    Returns whether the topology has a readable server available */
+static PHP_METHOD(TopologyDescription, hasReadableServer)
+{
+	php_phongo_topologydescription_t* intern;
+	const mongoc_read_prefs_t*        read_preference   = NULL;
+	zval*                             z_read_preference = NULL;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_START(0, 1)
+	Z_PARAM_OPTIONAL
+	Z_PARAM_OBJECT_OF_CLASS(z_read_preference, php_phongo_readpreference_ce)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	if (z_read_preference) {
+		read_preference = phongo_read_preference_from_zval(z_read_preference);
+	}
+
+	RETVAL_BOOL(mongoc_topology_description_has_readable_server(intern->topology_description, read_preference));
+} /* }}} */
+
+/* {{{ proto boolean MongoDB\Driver\TopologyDescription::hasWritableServer()
+    Returns whether the topology has a writable server available */
+static PHP_METHOD(TopologyDescription, hasWritableServer)
+{
+	php_phongo_topologydescription_t* intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_BOOL(mongoc_topology_description_has_writable_server(intern->topology_description));
+} /* }}} */
+
+/* {{{ proto string MongoDB\Driver\TopologyDescription::getType()
+    Returns the topology type */
+static PHP_METHOD(TopologyDescription, getType)
+{
+	php_phongo_topologydescription_t* intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_STRING(mongoc_topology_description_type(intern->topology_description));
+} /* }}} */
+
+/* {{{ MongoDB\Driver\TopologyDescription function entries */
+/* clang-format off */
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(ai_TopologyDescription_hasReadableServer, 0, 0, _IS_BOOL, 0)
```

Comment:
I saw that `ZEND_BEGIN_ARG_WITH_TENTATIVE_RETURN_TYPE_INFO_EX` is used in https://github.com/mongodb/mongo-php-driver/pull/1245 but for some reason the tests wouldn't compile when I used that instead of `ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX`. What is the difference between the two / what does it mean for the return type to be tentative?

Reply:
I believe `ZEND_BEGIN_ARG_WITH_TENTATIVE_RETURN_TYPE_INFO_EX` is the PHP internal equivalent of the `ReturnTypeWillChange` annotation for userland code (see: https://github.com/mongodb/mongo-php-library/pull/855). Both of these were introduced by https://wiki.php.net/rfc/internal_method_return_types and are intended to serve as a bridge in PHP 8.x until PHP 9.0 arrives. In the case of #1245, we needed to use them on our classes that implement core PHP interfaces, since those interfaces now declare a tentative return type.

As far as it being "tentative", I think this has to do with not being able to formally change the return type of interface methods since that would constitute a BC break (see: [Semantic Versioning and Public Interfaces](http://paul-m-jones.com/post/2015/06/02/semantic-versioning-and-public-interfaces/) for an explanation). Instead, code in PHP 8.1+ will need to start declaring a tentative return type or risk a deprecation warning (which is enough to make our own tests fail, which is why we addressed this).

In contrast to all of that, `ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX` is declaring the formal return type. I don't believe you actually need to worry about using this here, as we'll eventually get to it when cleaning up the entire extension in [PHPC-1709](https://jira.mongodb.org/browse/PHPC-1709). However, there's certainly no harm in doing so now if you can get this working.

The more important thing to address with respect to arginfo is the requirements for the ReadPreference parameter, as that's not something we could alter later without introducing a BC break.

Tying this back to the earlier blog post I referenced about semantic versioning and interfaces, we can effectively narrow the return type (e.g. anything -> bool) without causing a BC break, but we _cannot_ narrow the type of an accepted parameter (e.g. anything -> ReadPreference) without causing a BC break.

Reply:
Ok, that makes sense. I'm just using `ZEND_BEGIN_ARG_INFO_EX` then since I couldn't get `ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX` to work.